EXIT_SUCCESS return code declaration

ryzhovau opened the pull request:
Looks like EXIT_SUCCESS is undeclared:

```
mipsel-linux-uclibc-gcc -o obj/i2p.o i2p.cpp -c -O2 -pipe -mips32 -mtune=mips32 -fno-caller-saves -std=gnu++0x -D_GLIBCXX_USE_NANOSLEEP  -I/home/ryzhovau/Entware/openwrt_trunk/staging_dir/target-mipsel-linux-gnu/opt/include -I/home/ryzhovau/Entware/openwrt_trunk/staging_dir/target-mipsel-linux-gnu/include -I/opt/entware/toolchain-entware/include
i2p.cpp: In function 'int main(int, char**)':
i2p.cpp:15:9: error: 'EXIT_SUCCESS' was not declared in this scope
make[3]: *** [obj/i2p.o] Error 1
make[3]: Leaving directory `/home/ryzhovau/Entware/openwrt_trunk/build_dir/target-mipsel-linux-gnu/i2pd-0.1-20140610'
```

gcc 4.6.4 + uClibc 0.9.32 cross-compilation toolchain for MIPS embedded systems.

ryzhovau commented:
[Пoртирoвал](https://github.com/Entware/rtndev/commit/27796698eca1baad350aada5555809ce4ea380a9) i2pd в репoзитoрий сoфта для рoутерoв. Спасибo ещё раз за рабoту.

orignal commented:
Под openwrt соберется?

ryzhovau commented:
Да. Это готовый файл для билдрута OpenWrt, только пути установки поправить надо с `/opt` на `/usr`.

orignal commented:
Отлично. Если есть рабочая сборка то займусь уменьшением потребления памяти - надо ужать до 10 мегов.
Кстати вот тут https://github.com/PurpleI2P/i2pd/blob/master/SSUSession.h на MIPS-е проблем не возникало? Просто руки не дошли этот кусок переделать

orignal commented:
Можете об этом в твиттере написать, а то куча народу ждет не дожедется версии под роутеры?

ryzhovau commented:
> Кстати вот тут https://github.com/PurpleI2P/i2pd/blob/master/SSUSession.h на MIPS-е проблем не возникало?

Каких проблем? Я пока убедился в том, что открываются i2p адреса через WebConsole. Насколько понимаю, `Tunnel creation success rate: ~10%` связан с большим количеством [таймаутов](http://pastebin.com/b6vbH93m), возникающих из-за невысокой производительностью CPU роутеров.

> Можете об этом в твиттере написать, а то куча народу ждет не дожедется версии под роутеры?

[Не вопрос](https://twitter.com/ryzhov_al/status/574991913272606721).

orignal commented:
Проблемы в том что там этот unit32 не выровнен.  Интелы от этого подтормаживают, а спарки например грохаются. MIPS не знаю. В остальных местах это уже вычистил, отсалосось только тут и в NTCP.

ryzhovau commented:
>  Интелы от этого подтормаживают, а спарки например грохаются. MIPS не знаю. 

Отрытие ресурса что через WebConsole, что через proxy занимает примерно минуту. [Вот](http://pastebin.com/FfAbQmUa) некоторая статистика работы во время открытия i2p-ресурса. Радует то, что на 100% CPU занять довольно редко: некоторое время после старта i2pd и редкими пиками во время открытия i2p ресурсов.

orignal commented:
Минута это много, думаю что узлов в базе еще недостаточно, нормально это несколько секунд.
То что MIPS в отличие от ARM не грузит процессор на 100% при шифровании это хорошо. Осталось только потребление памяти.

ryzhovau commented:
> Минута это много, думаю что узлов в базе еще недостаточно,

После двеннадцати часов аптайма `Tunnel creation success rate: 42%`, ресурсы открываются за 30-60 секунд, таймаутов или сообщений `LeaseSet not found` больше нет. Потребление памяти или ресурсов не растёт. В логах довольно часто встречается `Unknown error 158`:
```
admin@RTN66U:/tmp/home/root# grep error /opt/var/log/i2pd.log | tail
10:13:09/error - No NTCP and SSU addresses available
10:13:09/error - No NTCP and SSU addresses available
10:13:19/error - Read error: Unknown error 158
10:13:20/error - SSU message has not been ACKed after 5 attempts. Deleted
10:13:33/error - Read error: Unknown error 158
10:13:35/error - SSU message has not been ACKed after 5 attempts. Deleted
10:13:50/error - SSU message has not been ACKed after 5 attempts. Deleted
10:13:53/error - SSU message has not been ACKed after 5 attempts. Deleted
10:13:57/error - Read error: Unknown error 158
10:13:58/error - Read error: Unknown error 158
```

orignal commented:
Сколько сейчас памяти ест? Что за ошибка 158 я посмотрю, возможно это просто "operation cancelled".

ryzhovau commented:
После 16-ти часов аптайма: 
```
admin@RTN66U:/tmp/home/root# pmap -d `pidof i2p` | grep shared
mapped: 45328K    writeable/private: 34240K    shared: 4K

admin@RTN66U:/tmp/home/root# du -hs /opt/var/i2pd/
8.7M    /opt/var/i2pd/
```
Отключил в дефолтной конфигурации пакета ведение логов, т.к. за полдня в лог-файл набежало 15МБ.

orignal commented:
15 Мб это не так много, у меня на флудфиле по нескольку гигов в день набегает. Хотя я все время количество записываемых данных уменьшаю. Потребление памяти все равно многовато, буду думать, где в первую очередь урезать. До вчерашнего дня данная задача была неактуальна.

ryzhovau commented:
[Записал](http://asciinema.org/a/17485) ascii cast для наглядности.

orignal commented:
Тот народ очень интересует какое железо

ryzhovau commented:
Asus RT-N66U: одноядерный MIPS @600MHz, 256MB RAM.

ryzhovau commented:
На NAS'е WD My Book Live (PowerPC, 800MHz, 256RAM, Debian Wheezy) ресурсы открываются примерно так же долго. После часа работы seg.fault'ов нет, но потребление памяти, естественно, больше, т.к. здесь полновесная glibc, а не uClibc как в случае с роутером:
```
(chroot-debian)~# pmap -d `pidof i2p` | grep shared
mapped: 134016K    writeable/private: 95552K    shared: 0K
(chroot-debian)~# uname -a
Linux MBL2TB 2.6.32.63-20140623-NCQ #1 Mon Jun 23 10:41:13 MSK 2014 ppc GNU/Linux
```

orignal commented:
Потребление памяти вызвано не библиотеками, а тем что через него больше данных пытаются гнать.

ryzhovau commented:
> Если есть рабочая сборка то займусь уменьшением потребления памяти - надо ужать до 10 мегов.

Попытался оценить что привнесли коммиты «reduce memory usage». Ниже представлен график изменения Private RAM в первые сто секунд после старта i2pd, коммиты dc599bbc636a7d0b290d953eaab37aedee3808f9 и f3aab569cab85ae0cd577aaf80ff8ff24e8178f5 соответствуют двум линиям old и new.

![RAM consumtion](http://s017.radikal.ru/i426/1503/5d/a91282ae8717.png)

На своих устройствах i2pd остановил и пока больше ничего писать здесь не собираюсь. Мне как мейнтейнеру репозитория достаточно того, что пакет рабочий. Но если нужно будет какое-то содействие — пожалуйста.

orignal commented:
Потреблением памяти при старте я еще не занимался - там вообщем то все понятно.
